### PR TITLE
[unitaryhack] Change the order of the covariance matrix and the vector of means internally in `default.gaussian` for `gaussian_state`

### PR DESCRIPTION
**Pull request by l45k**
**Context:**

Change the order of the covariance matrix and the vector of means internally in default.gaussian for gaussian_state

**Description of the Change:**

Changed the order of mu and cov in gaussian_state and adapted the documentation

**Benefits:**

Improved internal code quality for `default.gaussian`.

**Possible Drawbacks:**
N/A

**Related GitHub Issues:**
Closes PennyLaneAI#1257

### CODE REVIEW

**Comment by l45k** [comment]
Hey @josh146 and @antalszava , 
I am not completyl shure if changeing the order of mu and cov is enough. I searched for other parts of the code that need to be changed, however I found none. So please guide me into the right direction if I missed something.

**Comment by antalszava** [comment]
Hi @l45k, what we would like is to always have the covariance matrix first and the vector of means second. The signature of the `gaussian_state` function should be good as is, what we would like to change is the return statement: from `return mu, cov` we'd like to go to `return cov, mu`.

This applies to the entire `pennylane/devices/default_gaussian.py` file and related tests: we would always like the covariance matrix to comes first and the vector of means to come second. It will likely also include subtleties when each function is being used because the code for `default.gaussian` assumes at the moment that the vector of means is first and the covariance matrix is second.

Examples where a change is required:
* `fock_prob` function
* `displacement` function (the way of how a `state` is represented)
* `vacuum_state` function and other state preparation functions (the way of how a `state` is represented)
* `photon_number` function,
* etc.

Let us know if you have further questions. :slightly_smiling_face:

**Comment by MaldoAlberto** [comment]
For this problem my team and I saw that the situation of bad value conversion is found in this repository https://github.com/PennyLaneAI/pennylane-sf 
or is it necessarily in the Pennylane repository?

**Comment by antalszava** [comment]
Hi @MaldoAlberto, could you point me to which part of PennyLane-SF you've found?

The `default.gaussian` device is independent of PennyLane-SF. and we would like to switch to the convention that Straweberry Fields and PennyLane-SF use (cov mat first, vector of means second) within `default.gaussian`.

We did have previously an issue that was assumed to be coming from PennyLane-SF, but turned out that it was due to the documentation in this repo.

**Comment by MaldoAlberto** [comment]
@antalszava 

from the example code the problem is this command qml.GaussianState(mean_vector, cov_mat, wires=[0, 1]) the values that are passed are the r, V following this in the file simulator.py in the command line 111 you have

op = self._operation_map[operation](*sf_par)

that sf_par passes the r,V commands to the strawberryfields file ops.py, but the arguments must be V and r, then to add the condition to modify any method like GaussianState must be considered in the lines of code 98-104, in the case of else if we add a reverse it will work GaussianState with r and V respectively :

        # convert PennyLane parameter conventions to
        # Strawberry Fields conventions
        if operation == "CatState":
            sf_par = (par[0] * np.exp(par[1] * 1j), par[2])
        else:
            sf_pair = par
            sf_pair.reverse()

But if there are more methods, other conditions for the _operation_map method should be considered.

I hope my explanation is correct.

**Comment by antalszava** [comment]
The `qml.GaussianState(mean_vector, cov_mat, wires=[0, 1])` is incorrect, this is what the issue https://github.com/PennyLaneAI/pennylane-sf/issues/67 boiled down to. It should be `qml.GaussianState(cov_mat, mean_vector, wires=[0, 1])`. This is also what the current documentation for `GaussianState` reflects:
[`GaussianState(V, r, wires)`](https://pennylane.readthedocs.io/en/latest/code/api/pennylane.GaussianState.html) where `V` is the covariance matrix and `r` is the vector of means. This problem was solved with #1255. The original confusion came from the fact, that the documentation was incorrect.

Currently, the #1257 issue focuses only on the `default.gaussian` device. This device is not related to PennyLane-SF, but it's stand-alone and can execute operations correctly, including `qml.GaussianState(cov_mat, mean_vector, wires=[0, 1])`.

Internally, however, the `default.gaussian` device still uses the old order of the covariance matrix and vector of means. We would like this to be adjusted, so that everywhere, even in the internal code of `default.gaussian` we always have the cov mat coming before the vector of means for example in function signatures, when passing arguments, when returning these quantities, etc.

Hope this helps, let us know if you have further questions! :)

**Comment by l45k** [comment]
@antalszava  thank you for clearifying! I will change it accordingly.

**Comment by l45k** [comment]
Just to make sure. The intended bahvior is to have the cov_matrix first and the mean_vector second. Do we expect the same for the return values? For example the `gaussian_state`

```
def gaussian_state(cov, mu, hbar=2.0):
    r"""Returns a Gaussian state.

    This is simply a bare wrapper function,
    since the means vector and covariance matrix
    can be passed via the parameters unchanged.

    Note that both the means vector and covariance
    matrix should be in :math:`(\x_1,\dots, \x_N, \p_1, \dots, \p_N)`
    ordering.

    Args:
        cov (array): covariance matrix. Must be dimension :math:`2N\times 2N`,
            where N is the number of modes
        mu (array): vector means. Must be length-:math:`2N`,
            where N is the number of modes
        hbar (float): (default 2) the value of :math:`\hbar` in the commutation
            relation :math:`[\x,\p]=i\hbar`

    Returns:
        tuple: the mean and the covariance matrix of the Gaussian state
    """
    ...
    return mu, cov
```

returns the mean first and covariance second.  From my understanding this should be changed as well?

Thanks in advance!

**Comment by antalszava** [comment]
@l45k yep, that's right! The reason behind this change would be to eliminate any confusion that might arise from the difference of the ordering in the code, so indeed, the return statement would also have to be changed.

**Comment by l45k** [comment]
I commited the changes. However, when I run the tests I get a lot of Failed test. I didn't had the time to check why this happens, but I gues this might be due to some other code that needs to be adjustments to run with the changed parameters.

**Comment by mariaschuld** [comment]
Hey @l45k, I only had a superficial look at the tests, but the values for the gradients are far off the expected values (and it is not just an ordering issue), so there still seems to be some fundamental problem with the code changes! Let us know when you have questions, or if you solved the failing tests and are ready for a re-review.

**Comment by crazy4pi314** [comment]
Hey @l45k  and the PennyLane team! I wanted to let folks know I am extending the unitaryHACK deadline till the end of this week for PRs that were already in progress to give some time to iterate with maintainers. Let me know if you have questions, or if I can be of help in getting this issue closed!

**Comment by l45k** [comment]
Hey @mariaschuld and @antalszava, I finally manage to find the error. I missed to change the order in one case of the poly_quad_expectations. This is now corrected and all test finish successfully. So I would like to give it anothter try.

**Comment by antalszava** [comment]
Hi @l45k, that's great to hear! :slightly_smiling_face: Could you resolve the conflict in the `.github/CHANGELOG.md` file please?

It should involve the following locally:
1. Checking out the master branch: `git checkout master`.
2. Pulling the latest changes from there: `git pull`
3. Checking out this feature branch: `git checkout issue_1257`
4. Merging the master branch into the feature branch: `git merge master`
5. Merging should indicate that there are conflicts in the `.github/CHANGELOG.md` file. Those can be resolved by opening the file and comparing the difference between the `<<<<<<< issue_1257` and `>>>>>>> master` markers.

**Comment by l45k** [comment]
@antalszava sorry, I used the wrong origin when I tried to merge the newest changes. I hope it works now. Please let me know if there is anything else I can do.

File: pennylane/ops/qubit.py [hunk 1]
```diff
@@ -2961,75 +2961,6 @@ def diagonalizing_gates(self):
         return [QubitUnitary(self.eigendecomposition["eigvec"].conj().T, wires=list(self.wires))]
 
 
-class Projector(Observable):
```
Comment (antalszava) [inline pennylane/ops/qubit.py:2964]: How come this and parts of the `channel.py` file were removed? :thinking:

**Comment by antalszava** [comment]
Hi @l45k, it seems that multiple checks are failing due to parts of the codebase being removed (e.g., Projector operation).

**Comment by l45k** [comment]
I wasn't able to get the source right. Therfore, I reseted the branch to the current HEAD of the master branch. I don't know why this closed the PR. I implemented the changes again and the test should now run without errors. Can you reopen the PR or do I have to open a new one?

**Comment by antalszava** [comment]
Hi @l45k, it looks the formatting check failed, as some sources were not formatted. Could you format the `tests/devices/test_default_gaussian.py` file?

Also, there seems to be a conflict in the changelog. This can be resolved locally for example by merging `master` into this feature branch and then carefully modifying the changelog file.

**Comment by antalszava** [comment]
Hi @l45k, it looks the formatting is still off. Note that `black-21.6b0` should be used and the command would be `black -l 100 tests/devices/test_default_gaussian.py`.

**Comment by co9olguy** [comment]
Hi @l45k, just wanted to check the status of this contribution. Is it ready for the team to give another round of review? Have you taken into considerations the comment from @antalszava?

**Comment by l45k** [comment]
Hey @antalszava, sorry to bother you again. I think I will need some guidance for this one. I only changed the order of the arguments and nothing within the code. So this error should have been there already before I touched the code. Do you have a suggestion how to solve it?

File: .github/CHANGELOG.md [hunk 2]
```diff
@@ -522,6 +522,9 @@ Olivia Di Matteo, Ashish Panigrahi
 
 <h3>Improvements</h3>
 
+* Change the order of the covariance matrix and the vector of means internally
```
Comment (antalszava) [inline .github/CHANGELOG.md:525]: Hi @l45k, this item now is listed for v0.16.0 of PennyLane, which we have released recently.

Therefore, two small things would be needed:
* Moving this entry up in the file and adding it to the `<h3>Improvements</h3>` section of the new, v0.17.0 release (~line 8),
* Adding yourself to the list of contributors, as that is a separate, new release with new contributors. You can do that on ~line 27.

File: pennylane/devices/default_gaussian.py [hunk 3]
```diff
@@ -458,25 +458,25 @@ def gaussian_state(cov, mu, hbar=2.0):
...
         cov (array): covariance matrix to insert
+        mu (array): vector of means to insert
 
     Returns:
         tuple: contains the vector of means and covariance matrix.
     """
-    mu0 = state[0]
-    cov0 = state[1]
+    mu0 = state[1]
+    cov0 = state[0]
```
Comment (antalszava) [inline pennylane/devices/default_gaussian.py:479]:
```suggestion
    cov0 = state[0]
    mu0 = state[1]
```

File: pennylane/devices/default_gaussian.py [hunk 4]
```diff
@@ -512,7 +512,7 @@ def photon_number(mu, cov, params, hbar=2.0):
     # pylint: disable=unused-argument
     ex = (np.trace(cov) + mu.T @ mu) / (2 * hbar) - 1 / 2
     var = (np.trace(cov @ cov) + 2 * mu.T @ cov @ mu) / (2 * hbar ** 2) - 1 / 4
-    return ex, var
+    return var, ex
```
Comment (antalszava) [inline pennylane/devices/default_gaussian.py:515]: How come these are swapped? :thinking: These should be returned independently to the order of the covariance matrix and the vector of means.

File: pennylane/devices/default_gaussian.py [hunk 5]
```diff
@@ -598,15 +598,15 @@ def poly_quad_expectations(mu, cov, wires, device_wires, params, hbar=2.0):
     groenewald_correction = np.sum([np.linalg.det(hbar * A[:, m][n]) for m in modes for n in modes])
     var -= groenewald_correction
 
-    return ex, var
+    return var, ex
```
Comment (antalszava) [inline pennylane/devices/default_gaussian.py:601]: Similar question as for the mean photon case regarding the change of order here.

File: pennylane/devices/default_gaussian.py [hunk 6]
```diff
@@ -779,13 +779,13 @@ def expand(self, S, wires):
     def expval(self, observable, wires, par):
 
         if observable == "PolyXP":
-            mu, cov = self._state
-            ev, var = self._observable_map[observable](
-                mu, cov, wires, self.wires, par, hbar=self.hbar
+            cov, mu = self._state
+            var, ev = self._observable_map[observable](
```
Comment (antalszava) [inline pennylane/devices/default_gaussian.py:783]: Maybe the order of the expectation value and variance could be kept here and then we could leave them unchanged everywhere?

File: pennylane/devices/default_gaussian.py [hunk 7]
```diff
@@ -180,17 +180,17 @@ def displacement(state, wire, alpha, hbar=2):
     """Displacement in the phase space.
 
     Args:
-        state (tuple): contains means vector and covariance matrix
+        state (tuple): contains covariance matrix and means vector
         wire (int): wire that the displacement acts on
         alpha (float): complex displacement
 
     Returns:
         tuple: contains the vector of means and covariance matrix
```
Comment (antalszava) [inline pennylane/devices/default_gaussian.py:188]:
```suggestion
        tuple: contains the covariance matrix and the vector of means
```

File: pennylane/devices/default_gaussian.py [hunk 8]
```diff
@@ -458,25 +458,25 @@ def gaussian_state(cov, mu, hbar=2.0):
...
 
 
-def set_state(state, wire, mu, cov):
+def set_state(state, wire, cov, mu):
     r"""Inserts a single mode Gaussian into the
     state representation of the complete system.
 
     Args:
         state (tuple): contains means vector
             and covariance matrix of existing state
```
Comment (antalszava) [inline pennylane/devices/default_gaussian.py:470]:
```suggestion
        state (tuple): contains covariance matrix
            and means vector of existing state
```

File: pennylane/devices/default_gaussian.py [hunk 9]
```diff
@@ -728,9 +728,9 @@ def apply(self, operation, wires, par):
```
Comment (antalszava) [inline pennylane/devices/default_gaussian.py:728]: There seems to be a message on line 723-724 where the mention of cov and vector of means can be swapped, maybe good to adjust that too: (cannot make a suggestion)
```python
                    "GaussianState covariance matrix or means vector is "
                    "the incorrect size for the number of subsystems."
```

File: pennylane/devices/default_gaussian.py [hunk 10]
```diff
@@ -855,13 +855,13 @@ def reset(self):
...
-        r"""Returns the vector of means and the covariance matrix of the specified wires.
+        r"""Returns the covariance matrix and the vector of means of the specified wires.
 
         Args:
             wires (Wires): requested wires
 
         Returns:
-            tuple (means, cov): means is an array containing the vector of means,
+            tuple (cov, means): means is an array containing the vector of means,
             and cov is a square array containing the covariance matrix
```
Comment (antalszava) [inline pennylane/devices/default_gaussian.py:865]:
```suggestion
            tuple (cov, means): cov is a square array containing the covariance matrix,
            and means is an array containing the vector of means 
```

**Review by antalszava — CHANGES_REQUESTED** [review]
Hi @l45k, thanks for the update to the formatting! The code cov check should be fine, we could override that if need be as it seems that no additional tests would be required.

Had a review of the changes, overall it is looking great! :tada: I'm wondering about the change to the order of the expectation value and variance values, left comments at the related places.

Other changes required would be just adjusting the changelog and applying a couple of suggestions.

File: pennylane/devices/default_gaussian.py (hunk 6 above)
Comment (l45k) [reply pennylane/devices/default_gaussian.py:783]: I reset the order in the expectations back to ex,  var and therefore reversed it to the old order in `expval` and `var`.

**Comment by l45k** [comment]
Hey @antalszava I implemented the required changes. I hope it is now all right.

File: .github/CHANGELOG.md [hunk 11]
```diff
@@ -141,6 +141,9 @@
...
+  in default.gaussian for the gaussian_state. [(#1331)](https://github.com/PennyLaneAI/pennylane/pull/1331)
```
Comment (antalszava) [inline .github/CHANGELOG.md:145]:
```suggestion
* Change the order of the covariance matrix and the vector of means internally
  in `default.gaussian`. [(#1331)](https://github.com/PennyLaneAI/pennylane/pull/1331)
```

**Review by antalszava — APPROVED** [review]
Looks good to me! :1st_place_medal: :tada: 

Thank you @l45k for going through all suggestions! This will be very helpful when working with `default.gaussian` to avoid any confusion in the future. It can save a lot of headaches. :slightly_smiling_face:

**Pull request merged**